### PR TITLE
Refactor card classes to be pydantic

### PR DESCRIPTION
**Pull request by iTitus**

### CODE REVIEW

File: tests/whist/core/cards/test_card_container.py [hunk 1]
```diff
@@ -6,6 +6,29 @@
...
+        self.assertEqual(0, len(cc))
+
+    def test_full(self):
+        cc = CardContainer.full()
+        self.assertEqual(52, len(cc))
+
+    @staticmethod
+    def test_dict_no_crash():
+        cc = CardContainer(cards=[Card(suit=Suit.HEARTS, rank=Rank.NUM_2)])
+        cc.dict()
```
Comment (Segelzwerg) [inline tests/whist/core/cards/test_card_container.py:28]: Maybe we also could check if it is dictionary?

File: tests/whist/core/cards/test_hand.py [hunk 2]
```diff
@@ -6,24 +6,24 @@
...
 
     def contains_not_suit(self):
-        hand = Hand()
+        hand = Hand.empty()
         hand.add(Card(suit=Suit.CLUBS, rank=Rank.NUM_8))
         self.assertTrue(hand.contains_suit(Suit.HEARTS))
 
     def test_empty(self):
-        self.assertEqual(Hand(), Hand.empty())
+        self.assertEqual(Hand(cards=()), Hand.empty())
```
Comment (Segelzwerg) [inline tests/whist/core/cards/test_hand.py:23]: Do we have to check if the length is zero?

File: whist/core/cards/card_container.py [hunk 3]
```diff
@@ -1,60 +1,117 @@
...
-            raise KeyError(f'{card} already in set')
-        self._cards.add(card)
+        if not isinstance(card, Card):
+            raise ValueError(f'cannot add {card} of type {type(card)} to card container')
+        if card in self._cards_set:
+            raise ValueError(f'{card} already in card container')
+        self._cards_set.add(card)
+        self.__resync()
+
+    def __resync(self):
```
Comment (Segelzwerg) [inline whist/core/cards/card_container.py:114]: Although it is private, can you give a short explaination in the documentation?

**Review by Segelzwerg — CHANGES_REQUESTED** [review]

**Comment by Segelzwerg** [comment]
Sorry for the last comments. I think the look so weird because I commented on the commit and not the pr - was using mobile app

File: tests/whist/core/cards/test_hand.py (hunk 2 above)
Comment (iTitus) [reply tests/whist/core/cards/test_hand.py]: i dont know? this is your test :P

File: whist/core/cards/card_container.py [hunk 4]
```diff
@@ -1,60 +1,120 @@
...
         """
         Remove a card from this container.
 
         :param card: card to remove
         """
-        self._cards.remove(card)
+        if not isinstance(card, Card):
+            raise ValueError(f'cannot remove {card} of type {type(card)} from card container')
+        if card not in self._cards_set:
+            raise ValueError(f'{card} not in card container')
```
Comment (Segelzwerg) [inline whist/core/cards/card_container.py:97]: Maybe we should use customs errors here.

File: whist/core/cards/card_container.py [hunk 5]
```diff
@@ -1,60 +1,120 @@
...
+        return f'CardContainer(cards={self.cards!r})'
 
     def remove(self, card: Card) -> None:
         """
         Remove a card from this container.
 
         :param card: card to remove
         """
-        self._cards.remove(card)
+        if not isinstance(card, Card):
```
Comment (Segelzwerg) [inline whist/core/cards/card_container.py:94]: Is that even possible?

File: whist/core/cards/card_container.py [hunk 6]
```diff
@@ -1,60 +1,120 @@
...
         """
         Add a card to this container.
 
         :param card: card to add
         """
-        if card in self._cards:
-            raise KeyError(f'{card} already in set')
-        self._cards.add(card)
+        if not isinstance(card, Card):
+            raise ValueError(f'cannot add {card} of type {type(card)} to card container')
```
Comment (Segelzwerg) [inline whist/core/cards/card_container.py:108]: This could also be a custom error?

File: whist/core/cards/card_container.py (hunk 4 above)
Comment (iTitus) [reply whist/core/cards/card_container.py:92]: ehh, no real need. I am using standard Python conventions here. Problems with arguments always raise `ValueError`

File: whist/core/cards/card_container.py (hunk 5 above)
Comment (iTitus) [reply whist/core/cards/card_container.py:89]: yeah of course, there's a test for it

File: whist/core/cards/card.py [hunk 7]
```diff
@@ -1,11 +1,13 @@
...
 
 from enum import Enum
-from typing import Any, Optional
+from functools import total_ordering
+from typing import Any, Optional, Iterator
 
 from pydantic import BaseModel
 
 
+@total_ordering
```
Comment (Segelzwerg) [inline whist/core/cards/card.py:10]: Is this replacing the __ge__ and others?

File: whist/core/cards/card_container.py (hunk 4 above)
Comment (Segelzwerg) [reply whist/core/cards/card_container.py:92]: Agreed, but it is much more convenient during exception handling to use custom errors. But we two different message here, I'm fine with it and it is up to you.

File: whist/core/cards/card_container.py [hunk 8]
```diff
@@ -1,60 +1,224 @@
...
-        self._cards.remove(card)
+        if not isinstance(card, Card):
+            raise ValueError(f'cannot remove {card} of type {type(card)} from card container')
+        if card not in self:
+            raise ValueError(f'{card} not in card container')
+        self._remove_impl(card)
+
+    @abc.abstractmethod
+    def _remove_impl(self, card: Card) -> None:
+        pass
```
Comment (Segelzwerg) [inline whist/core/cards/card_container.py:97]: Shouldn't we raise `NotImplementedError` here?

**Comment by Segelzwerg** [comment]
I run tests against the main branch in server and we had no issues there. After resolving the above I'll approve.

File: whist/core/cards/card.py (hunk 7 above)
Comment (iTitus) [reply whist/core/cards/card.py:10]: yes. i only have to implement __eq__ and one comparison and then everything else is added automatically

File: whist/core/cards/card_container.py (hunk 4 above)
Comment (iTitus) [reply whist/core/cards/card_container.py:92]: ok

File: whist/core/cards/card_container.py (hunk 8 above)
Comment (iTitus) [reply whist/core/cards/card_container.py:97]: it does not matter. this method does not really exist and will never be called. i will change it

**Review by Segelzwerg — APPROVED** [review]

**Pull request merged**